Documentation/bpf.rst

```diff
@@ -3354,7 +3354,376 @@ the toolchain section under the respective tools.
+  by itself in an efficient manner and thus there is no need for anything other than
+  ``direct-action`` mode.
+
+..
+
+* **Question:** Is there any performance difference in offloaded ``cls_bpf`` and
+  offloaded XDP?
+* **Answer:** No. Both are JITed through the same compiler in the kernel which
+  handles the offloading to the SmartNIC and the loading mechanism for both is
+  very similar as well. Thus, the BPF program gets translated into the same target
+  instruction set in order to be able to run on the NIC natively. The two tc BPF
+  and XDP BPF program types have a differing set of features, so depending on the
+  use case one might be picked over the other due to availability of certain helper
+  functions in the offload case, for example.
+
+**Use cases for tc BPF**
+
+Some of the main use cases for tc BPF programs are presented in this subsection.
+Also here, the list is non-exhaustive and given the programmability and efficiency
+of tc BPF, it can easily be tailored and integrated into orchestration systems
+in order to solve very specific use cases. While some use cases with XDP may overlap,
+tc BPF and XDP BPF are mostly complementary to each other and both can also be
+used at the same time or one over the other depending which is most suitable for a
+given problem to solve.
+
+* **Policy enforcement for containers**
+
+  One application which tc BPF programs are suitable for is to implement policy
+  enforcement, custom firewalling or similar security measures for containers or
+  pods, respectively. In the conventional case, container isolation is implemented
+  through network namespaces with veth networking devices connecting the host's
+  initial namespace with the dedicated container's namespace. Since one end of
+  the veth pair has been moved into the container's namespace whereas the other
+  end remains in the initial namespace of the host, all network traffic from the
+  container has to pass through the host-facing veth device allowing for attaching
+  tc BPF programs on the tc ingress and egress hook of the veth. Network traffic
+  going into the container will pass through the host-facing veth's tc egress
+  hook whereas network traffic coming from the container will pass through the
+  host-facing veth's tc ingress hook.
+
+  For virtual devices like veth devices XDP is unsuitable in this case since the
+  kernel operates solely on a ``skb`` here and generic XDP has a few limitations
+  where it does not operate with cloned ``skb``'s. The latter is heavily used
+  from the TCP/IP stack in order to hold data segments for retransmission where
+  the generic XDP hook would simply get bypassed instead. Moreover, generic XDP
+  needs to linearize the entire ``skb`` resulting in heavily degraded performance.
+  tc BPF on the other hand is more flexible as it specializes on the ``skb``
+  input context case and thus does not need to cope with the limitations from
+  generic XDP.
```

Comment:
Is the intended audience once that knows what an `skb` is? we use `sk_buff` everywhere else, it seems?

Reply:
Yes, they should know.